### ISSUE TITLE
Better error messages from the parser and code analyser

### ISSUE THREAD

**Issue by fmontesi** [jolie/jolie#108]
Hi @jolie/developers,

Our current error reporting system works, but is not very user friendly. Consider this program:
```
main
{
	x = 3;
}
```

The extra semi-colon at the end is going to make `jolie --check` output:

```
jolie.lang.parse.ParserException: /home/fmontesi/tmp/main.ol:5: error: expected basic statement. Found token type RCURLY
        at jolie.lang.parse.AbstractParser.throwException(AbstractParser.java:241)
        at jolie.lang.parse.OLParser.parseBasicStatement(OLParser.java:1889)
        at jolie.lang.parse.OLParser.parseSequenceStatement(OLParser.java:1535)
        at jolie.lang.parse.OLParser.parseParallelStatement(OLParser.java:1518)
        at jolie.lang.parse.OLParser.parseProcess(OLParser.java:1511)
        at jolie.lang.parse.OLParser.parseMain(OLParser.java:1453)
        at jolie.lang.parse.OLParser.parseCode(OLParser.java:1434)
        at jolie.lang.parse.OLParser._parse(OLParser.java:219)
        at jolie.lang.parse.OLParser.parse(OLParser.java:185)
        at jolie.Interpreter.buildOOIT(Interpreter.java:1236)
        at jolie.Interpreter.init(Interpreter.java:1038)
        at jolie.Interpreter.run(Interpreter.java:1124)
        at jolie.Jolie.main(Jolie.java:60)
```

# Problem 1

The stack trace of the Java exception is just scary and distracting. It should be optional, and the default should be not to show it. This suggests that we should have a few flags to govern the format of error reporting. We should still be able to ask people to give us the stack trace when we need it, since sometimes it is useful for finding bugs.

# Problem 2

The error message `expected basic statement` does not give an idea of how to fix the bug. It is too technical. We should pair errors with suggestions. For example, here it would be nice to tell the programmer that she probably just inserted an extra semi-colon.

# Problem 3

Our Atom IDE supports reporting lots of interesting stuff, like suggestions and links to webpages that explain errors in more detail. We are not currently using these advanced features because we do not get anything to report that is interesting enough from `jolie --check`. For example, check this issue: https://github.com/fmontesi/linter-jolie/issues/7

If we want to report a lot of stuff, it should be structured. We cannot just continue living with a string. I suggest to add a flag that allows us to ask Jolie to output errors in a given format. For example, `jolie --check --check-output-format=json` (`--check-output-format` is too long, it's just an example; please suggest a better alternative). This would then output the error in json format. Here's an example with the previous piece of code:

```
{
    location: {
        file: "/home/fmontesi/tmp/main.ol",
        position: [ [3,7], [3,8] ]
    },    
    error: "Expected basic statement. Found }.",
    description: "You may have an extra semi-colon",
    url: "http://docs.jolie-lang.org/#!documentation/basics/composing_statements.html",
}
```

@agwe: This could be a great beginner project.

@jolie/developers: If you know of anybody who would like to contribute, please notify me.

**Comment by bmaschio** [comment]
@jolie/developers,
Problem 1 
We could have a --verbose parameters for "nerds" that like java style dump
Problem 3
I really like the idea of  json this will open the possibility of  developing web base developing tools 

How can I help ?

**Comment by fmontesi** [comment]
Hi Balint,

I think that a good starting point is proposing the flags for controlling the output, like getting the "verbose" error messages or the json format. Then, we need to update the parser such that it also returns column positions in its exceptions. Once that is done, we can start working on the JSON output.

**Comment by oxxford** [comment]
Here is a list of all error messages. Choose, please, which of them should be replaced:
https://docs.google.com/document/d/1YDgByoVaUIEDqXhRMRjfuIfkBGIL2RsNvfz91fEkrZU/edit?usp=sharing

**Comment by bmaschio** [comment]
@fmontesi  @oxxford 

Hi 

I had a look at the list some and I have put down down some of my suggestions.

In general my way of thinking is that we should try to give the user more "error localization " information possible. For example if the error is on a Outport port definition it would be nice to have something like 

> Syntax error at [Name of the outport] definition

or in the case of Type Definition  Error 

> missing dot in node definition for type [Name of type]

What do you think ?

Regards

Balint

**Comment by fmontesi** [comment]
@oxxford: Thanks for the doc! 
@bmaschio: definitely.

@oxxford and @bmaschio : Please try to think about suggestions/descriptions too when we encounter an error. I'm not 100% convinced we can give useful suggestions yet, but it'd be nice to see that on a couple of examples.

**Comment by oxxford** [comment]
@fmontesi 
I'm a bit confused about what exact suggestions should we provide. Isn't (for example) report "expected ;" comprehensive enough. For example, Intelij idea use similar reports. Could you please give a better example?

**Comment by klag** [comment]
I think we should enhance the description in order to give hints to the developer to better understand the new programming paradigm. Think at the errors as a source of documentation. For example, in the classical case of a ; missing we usually get errors like:
: error: expected }. Found token type ID, token content...

Such a message is quite difficult to be intepreted by a newbie. Maybe something like:

Syntax error: expected }. Found token type ID, token content ....
[HELP] Check if a ; is missing. In Jolie ; is an operator of sequence not an ending line symbol. 

Another example (in the case I develop an operation XXX in the body which is not present in the defined interface)
Semantic Error: invalid identifier: XXX
[HELP] You attempt to define an operation in the body which is not declared in the inputPorts of this service. Please check the related inputPort interfaces or, maybe, fix the name of the operation.

And so on...

**Comment by bmaschio** [comment]
@klag 
I agree with you on the need of explaining better the paradigm. I would not suggest to have it as output on the command screen.  This will force us to update the Jolie source code every time that we want to add more inclusive explanations. I world take the approach of refering the programmer to a page on the site so that we can start with a simple explanation and building up. Also it will be good because Google will be indexing the pages so i suppose if a programmer would search " jolie error semicolon " something will come out

**Comment by klag** [comment]
I think we could have both. A brief hint from command line and an url to follow in order to get a more exhaustive explanation

**Comment by oxxford** [comment]
I added suggestions you gave (they are marked green in the list) and did some my own suggestions (they are marked black in the list). Could you check if they are good enough to add it to the code ( you can mark it yellow if they satisfy your expectations)?

**Comment by bmaschio** [comment]
@fmontesi  shell I pick up this

**Comment by fmontesi** [comment]
OK for me, but what about @oxxford . @lsafina what's the status on this?

**Comment by bmaschio** [comment]
@fmontesi @jolie/developers 

Hello everyone

I did an early attempt
I added  --outputType to the command parser possibilities 

**jolie --check  --outputType json main.ol**

with the following result

ott 29, 2017 7:50:35 PM jolie.lang.parse.SemanticVerifier validate
GRAVE: Aborting: input file semantically invalid.
ott 29, 2017 7:50:35 PM jolie.Interpreter buildOOIT
GRAVE: {"filename":"C:\\TestJolie\\main.ol",
                "line":5,
               "error":"Call to undefined definition: dasdada"
             }

this has been  implemented by taking down to the SemanticError class the parameter errorJSON as a boolean

My idea would proceed in this way

1. Remove the "Aborting: input file semantically invalid."

2. Add that information at the json   {"errorType":"SemanticError",
                                                            "filename":"C:\\TestJolie\\main.ol",
                                                            "line":5,
                                                            "error":"Call to undefined definition: dasdada"
                                                           }
3. Adding a error code to the SemmanticError class to handle the call to jolie help web page  

Any comments

Balint

**Comment by fmontesi** [comment]
Hi Balint,

Thanks for starting this! This kinda inspired me: shouldn't we make *all* log messages configurable like this? Currently all log messages (errors, warnings, etc.) by the interpreter are output by using a few internal loggers or the interpreter logging methods like Interpreter.logWarning.

So maybe we could:
- Make a JolieLogger interface with the methods we need, like logWarning.
- When we launch the interpreter, we set the right class implementing JolieLogger in the Interpreter object (we will immediately have two possibilities: PlainConsoleLogger that does what we do now, and JSONConsoleLogger that prints everything in JSON format). We decide which implementation of JolieLogger we want by looking at the command line.
- Make sure that every time the interpreter wants to log something, we use the JolieLogger object stored in the Interpreter.

@jolie/developers what do you all think?

Cheers,
Fabrizio

**Comment by klag** [comment]
I think it sounds good. I also like the idea to list all the available errors as in the attempt of oxxford.

**Comment by thesave** [comment]
Sounds good also to me and in line with the current effort to output `--trace`s in convenient formats (e.g.,  project [jtracer](https://github.com/VincenzoMattarella/jolieJTracer) and its [companion visualiser](https://github.com/VincenzoMattarella/jtracer-1) by @VincenzoMattarella, soon to be merged within the official Jolie repos).

**Comment by fmontesi** [comment]
Yes. My vision is that everything will converge into a single logging
facility, for both tracing and error messages, where we just use tags to
distinguish them. Even monitoring could converge in all this!

Cheers,
F

**Comment by bmaschio** [comment]
@fmontesi @jolie/developers  
I had a look at some stuff 
In the Interpreter class, a standard logger (log4j) the different level of logging are rapped by some public methods

1. public void logSevere( String message ) => logger.severe( logPrefix + message );
2. public void logSevere( Throwable t ) => for logging Throwable.

Now I imagine that @fmontesi  idea was to substitute the logger object  with a logger class that implements a specific interface that could look like something like this 

 public interface LoggerInterface {
  public void logSevere(String message);
  public void logSevere(Throwable t);
  public void logSevere(LoggerMessage l);
  public void logWarning(LoggerMessage l);
  public void logWarning(String message);
  public void logInfo(LoggerMessage l);
  public void logInfo(String message);
}
What do you think about it?

@thesave 
does the jtracer has impact on the interpreter?

Cheers
Balint

**Comment by fmontesi** [comment]
Hi Balint,

Precisely!

To discuss the API that our Logger should have, I think we need an example for each each "class" of logs. Or, in other words, we need to figure out what's inside a LoggerMessage as you wrote in your API.

I can imagine at least a few classes of messages:
- Uncaught faults;
- Runtime interpreter errors (e.g., problems with sockets, out of memory errors, etc.);
- Monitoring/tracing events (e.g., messages sent or received);
- Code analysis messages (e.g., linting warnings, syntax errors, etc.).

Anything else?

Cheers,
F

**Comment by thesave** [comment]
@bmaschio yep, to support logging in JSON @VincenzoMattarella had to modify a few classes of the interpreter, e.g., to read and enable logging from console commands and to route and convert log messages - @VincenzoMattarella feel free to correct me or add details if necessary.

As you say, the next step for jtracer is to become an optional module hooked to the logging API provided by the interpreter.

The classes listed by @fmontesi seem comprehensive.

**Comment by bmaschio** [comment]
@thesave 
I had a look at @VincenzoMattarella  good work and I have few questions on how to proceed 
in @VincenzoMattarella work the CommandParser  
```
           else if ("--trace".equals(argsList.get(i))) {
                optionsList.add(argsList.get(i));
                bTracer = true;
            } else if ("--trace:json".equals(argsList.get(i))) {
                optionsList.add(argsList.get(i));
                bJTracer = true; 
```
Fondamentally he defines a new options for the --trace command --trace:json
If we want to proceed this way i think it may result to a prolifaration of command options ex

1. --trace:xml
2. --trace:normal
3. --check:json
4. --check:normal
5. --check:outputType

I have two proposition 
1. going for the format --optionsName outputType  --trace json
2  adding --outputFomat outputType

The first option will require to specifying each time output type, the second will allow to a have a standard type if the output type is not specified plus it can be applied the outputType to the Jolie exception. @thesave @VincenzoMattarella @fmontesi what do you think?

@fmontesi  

This an idea for LoggerMessage 
```
public class LoggerMessage {
    private String fileName;
    private String message;
    private Value  data;
    
   public void LoggerMessage(){}
   public void LoggerMessage(String fileName, String message ){
      this.fileName = fileName;
      this.message = message;
   } 
   
    public void LoggerMessage(String fileName, String message , Value data ){
      this.fileName = fileName;
      this.message = message;
      this.data = data;
   } 
    
   public String getFileName(){
      return fileName ;   
       } 
    public String getMessage(){
      return message ;   
       } 
   public Value getData(){
      return data;
   }
   
}
```
What you think

Balint

**Comment by fmontesi** [comment]
Hi @bmaschio 

# Ad the command line options

I don't have a specific proposal in mind yet, but I would try to see if we can centralise everything in a single command (or two, one for output format and the other for what to log).
Here's a possible syntax: `--log:<FORMAT>:[list of WHAT]`
Here's an example: `--log:json:trace,faults`

One could then use it multiple times if different formats are wished for, for example: `--log:json:trace --log:text:faults --log:xml:runtime`

@jolie/developers Do we need this flexibility? Or, would we rather have that all messages have the same format? In that case we may have two flags, one like `--logFormat json` to determine the format, and another like `--log json,trace,faults` to tell Jolie what should be logged.

# Ad LoggerMessage

- I don't see why you need filename: is it shared by all logs?
- What is message? Why is it not inside of the data?
- Is the type of the message important? Meaning: do we care to mention it explicitly at the top level of the output somehow that the log is for a fault, a trace event, or a linter message? If so, we may need a LoggerMessage.Type enumeration to indicate which kind of message it is.

Cheers,
Fabrizio

**Comment by bmaschio** [comment]
@fmontesi 

- Yes but it derives from the context.sourceName() maybe I am wrong but would be not better to abstract as much as possible, see it as LoggerMessage Emitter  instead of  filename

- Message is the code errorCode E001 that can be used to then compose http://jolielang.help.com/ErrorCode= 10

- Perhaps no all messages have data?

- Yes we should LoggerMessage.Type enumeration very good idea

Cheers
Balint

**Comment by fmontesi** [comment]
@bmaschio 

1. OK, so we could have a `String source()` method. This sounds related to the `logPrefix` field used by the interpreter. Maybe that's the source, and we don't need one in the LoggerMessage interface, it just gets added by the interpreter in the output log. Thus I'm not entirely convinced we need one yet.
2. Mmh, so you want that *all* log messages to have a link that points to an explanation of the message. This makes sense to me. @jolie/developers what do you think? Does it make sense for *all* log messages? If you think not, please mention an example. And what could a default link for logs without a specific link be?

**Comment by bmaschio** [comment]
@fmontesi 
Sorry, @fmontesi  I have been super busy with work I have created some classes to implement some of my ideas.
 

- LogMessage contains the logMessage 
- AbstractLogger this is an abstract class that will allow implementing our specific logger classes.
- StandardLogger implements writeLogger as we like
- AbstractLogger has method like pushLog() flushSevere()  flushWarning() flushTrace() that allow to use  your log as list to be flush

You can find in my branch of Jolie

Cheers 

Balint

**Comment by fmontesi** [comment]
Hi Balint,

- Why should we have the lists and flushes?
- Why is there the loggerLevel field in the logger? It seems unused.

Cheers,
F

**Comment by bmaschio** [comment]
@fmontesi 

- we could think to have a single point of printing at the top level and play with the tabs or json subdocument to give a sense of chain of events

- Do you mean in the loggerMessage?

Cheers
B

**Comment by bmaschio** [comment]
@fmontesi

{"Level":"SEVERE","ErrorCode":"ParsingError","data":{"line":"12","sourceName":"C:\TestJolie\main.ol","message":"Communication protocol extension for protocol sodep not found."}}	

how does it look?

**Comment by fmontesi** [comment]
Very good. :+1: 

In your example, the usefulness of the error code is unclear. Maybe we should remove it, since a generic "ParsingError" web page would not be very helpful. I would keep lowercase letters for all fields, for consistency.

```
{
  "level":"severe",
  "data": {
    "line":"12",
    "sourceName":"C:\TestJolie\main.ol",
    "message":"Communication protocol extension for protocol sodep not found."
  }
}
```

**Comment by fmontesi** [comment]
Very good. :+1: 

In your example, the usefulness of the error code is unclear. Maybe we should remove it, since a generic "ParsingError" web page would not be very helpful. However, "ParsingError" may make sense as a type. The one you have in there is not really a "parsing" error though (it's not related to parsing), we need to find a better name (`ProgramError` ? not too happy about that either... suggestions, @jolie/developers ?). I would keep lowercase letters for all fields, for consistency.

```
{
  "level":"severe",
  "type":"ProgramError",
  "data": {
    "line":"12",
    "sourceName":"C:\TestJolie\main.ol",
    "message":"Communication protocol extension for protocol sodep not found."
  }
}
```

**Comment by bmaschio** [comment]
@fmontesi 
My implementation has intercepted the error on the 'catch' part of the code when the exception has already been fired. Perhaps we could consider adding a LogMessage to each exception so to be more near to the source of the error.
```
{
  "level":"severe",
  "type":"ProgramError",
   "errorCode": "code"
  "data": {
    "line":"12",
    "sourceName":"C:\TestJolie\main.ol",
    "message":"Communication protocol extension for protocol sodep not found."
  }
}
```   
What you think

B

**Comment by fmontesi** [comment]
Mmh I don't understand. Can you show me an example of what a code might be?

- F

**Comment by bmaschio** [comment]
@fmontesi 
Here an example of my implementation at AggregatedOperation.java  

```  
catch( TypeCheckingException e ) {
                            
             LogMessage logMessage = new LogMessage();
             Value v = Value.create();                
             v.getNewChild("message").setValue("Received message TypeMismatch (inputoperation \"" + operation.id() + "): " + e.getMessage());
              v.getNewChild("operationId").setValue(operation.id());
              logMessage.setErrorType("TypeMismatch");
	      interpreter.logWarning( logMessage );
```  
The value contained in LogMessage is limitated to the information that i can retrive from 'TypeCheckingException' 
Now if 'TypeCheckingException'  would implement the following interface
```  
public interface ExeptionLoggerInterface {
    public void setMessage(LogMessage l);
    public LogMessage getMessage();
}
```  
we could enrich the 'TypeCheckingException'  with details like  an errorCode each time that used   

Cheers
Balint

**Comment by mwallnoefer** [comment]
I'm going to close this issue for now since I guess we had some improvements meanwhile. If necessary, it is better to open a new issue with the outstanding tasks.
